Fix test_distributions when WITH_SCALARS.

gchanan opened the pull request.

gchanan committed:
Fix test_distributions when WITH_SCALARS.

fritzo commented on torch/distributions/categorical.py:
```diff
@@ -73,11 +73,19 @@ def param_shape(self):
-        return self.probs.new([float('nan')]).expand(self._extended_shape())
+        # verbose because of differening Variable/Tensor apis and lack of dtypes
+        if isinstance(self.probs, Variable):
+            return variable(float('nan')).type_as(self.probs).expand(self._extended_shape())
```
Does `.type_as()` place its result on the correct cuda device?

gchanan replied:
apparently not.

fritzo replied:
I always use `probs.new([float('nan')]).expand(...)`. What is the new syntax we can use for scalars to get correct device placement?

fritzo commented on torch/distributions/pareto.py:
```diff
@@ -29,7 +29,7 @@ class Pareto(TransformedDistribution):
     def __init__(self, scale, alpha):
         self.scale, self.alpha = broadcast_all(scale, alpha)
-        base_dist = Exponential(alpha)
+        base_dist = Exponential(self.alpha)
```
cc @alicanb

gchanan committed:
Use SCALAR_SHAPE in test, use self.scale in AffineTransform.

gchanan commented:
ok this works now both with and without scalars; I'll think of a workaround for the type_as issue.

fritzo commented on torch/distributions/pareto.py:
```diff
@@ -29,8 +29,8 @@ class Pareto(TransformedDistribution):
-        base_dist = Exponential(alpha)
-        transforms = [ExpTransform(), AffineTransform(loc=0, scale=scale)]
+        base_dist = Exponential(self.alpha)
+        transforms = [ExpTransform(), AffineTransform(loc=0, scale=self.scale)]
```
@alicanb pointed out that if `scale` and `alpha` are both scalars, then the 0 here will broadcast everything up to dimension-1. Does it seem reasonable to use `loc=variable(0.0)` here instead, at least until Python numbers become 0-dimensional?

alicanb replied:
`self.scale` is already the output of `broadcast_all` so I think it's safe now

fritzo replied:
but until `WITH_SCALARS=1`, 0 may be a higher-dimensional tensor than `self.scale` and therefore this broadcasting is wrong

gchanan replied:
Won't self.scale be at least 1-d?

fritzo replied:
Oh maybe I misunderstood. Is the following true?
- If `WITH_SCALARS=0` then tensors have at least 1 dim and python numbers have 1 dim
- If `WITH_SCALARS=1` then tensors may have 0 dims, and python numbers have 0 dims

If so then @alicanb is right and this is fine.

gchanan replied:
That's right if by "python numbers have X dim" you mean python numbers get broadcasted to a tensor of X dims in broadcast_all; python numbers themselves don't have dims.

gchanan committed:
Handle device correctly for scalars.

gchanan commented:
@pytorchbot retest this please.

gchanan committed:
Merge remote-tracking branch 'origin/master' into fix_dist_shapes3

gchanan committed:
Fix one hot categorical.

fritzo commented:
Here's a fix for `OneHotCategorical`:
```diff
diff --git a/torch/distributions/relaxed_categorical.py b/torch/distributions/relaxed_categorical.py
index 1961aeb0..54538f1d 100644
--- a/torch/distributions/relaxed_categorical.py
+++ b/torch/distributions/relaxed_categorical.py
@@ -65,7 +65,7 @@ class ExpRelaxedCategorical(Distribution):
     K = self._categorical._num_events
     self._validate_log_prob_arg(value)
     logits, value = broadcast_all(self.logits, value)
-    log_scale = (self.logits.new(1).fill_(K).lgamma() -
+    log_scale = (self.temperature.new(self.temperature.shape).fill_(K).lgamma() -
           self.temperature.log().mul(-(K - 1)))
     score = logits - value.mul(self.temperature)
     score = (score - log_sum_exp(score)).sum(-1)
```

gchanan committed:
Add a new_tensor instance method to Variable that takes only data.

This is to work around the legacy problems of new, where e.g.
new(5) will give you an unfilled tensor rather than a scalar.

gchanan committed:
Fix relaxed categorical.

gchanan committed:
Remove double return.

gchanan committed:
Fix cuda scalar code path.

gchanan committed:
Work around lack of WITH_SCALARS.

gchanan committed:
Use tensor_new.

gchanan commented:
This now includes https://github.com/pytorch/pytorch/pull/5144.

gchanan commented:
@pytorchbot retest this please.
